src/forward/compiler.jl

```diff
@@ -70,7 +70,91 @@ function dual(ir)
   return ir
 end
 
-@dynamo function _pushforward(_, x...)
+const chainrules_frule_fallback = which(frule, Tuple{Any})
+
+"""
+  has_chain_frule(T)
+
+For a type-tuple `T` e.g. `Tuple{typeof(f), Int, Float64}`, checks if there is a `rrule` defined for it.
+Excluding the generic fallback.
+The first return value is `true` if the `rrule` exists, `false` otherwise.
+If it does not, then the second argument is a list of edges to attach to the CodeInfo for a generated function,
+such that if a suitable rule is defined later, the generated function will recompile.
+"""
+function has_chain_frule(T)
+  m = meta(Tuple{typeof(frule), T.parameters...})
+
+  if m.method !== chainrules_frule_fallback
+    # found a frule, no need to add any edges
+    return true, nothing
+  end
+
+  return false, m.instance
+end
+
+"""
+    chain_frule(f, args...)
+
+Returns a the (primal) value of `f(args...)` and tangent, by invoking
+`ChainRules.frule(f, args...)`.
+"""
+@noinline chain_frule(dargs, args...) = frule(dargs, args...)
+
+"""
+  chain_frule_kw(kwf, kwargs, f, args...)
+
+As per [`chain_frule`](@ref) but with support for kwargs.
+`kwf` should be the kwfunc matching to `f`, and `kwargs` are a `NamedTuple` of keyword
+arguments.
+"""
+@inline chain_frule_kw(kwf, kwargs, f, args...) = frule(f, args...; kwargs...)
```

Comment:
This doesn't seem right.
`frule` takes `dargs` first before `f`.
Maybe:

```suggestion
@inline chain_frule_kw(kwf, kwargs, dargs, f, args...) = frule(dargs, f, args...; kwargs...)
```
which acts the same under most circumstances

Reply:
Yup, I just haven't done anything with kwargs yet, thanks for pointing this out though.